upath/implementations/local.py

```diff
@@ -0,0 +1,92 @@
+from __future__ import annotations
+
+import os
+from pathlib import Path
+from pathlib import PosixPath
+from pathlib import WindowsPath
+from typing import Any
+from typing import Iterable
+
+from fsspec.implementations.local import LocalFileSystem
+
+from upath.core import UPath
+
+__all__ = [
+    "LocalPath",
+    "PosixUPath",
+    "WindowsUPath",
+]
+
+
+class LocalPath(UPath):
```

Comment:
What is `LocalPath` for? Should `WindowsUPath` and `PosixUPath` subclass `LocalPath`?

Reply:
LocalPath handles the "file://" uri style LocalPaths.

The minor difference between the two is:
- LocalPath provides local file access through fsspec.
- WindowsUPath and PosixUPath via pathlib.

For your usecase do you just need to determine if a path is local, or if the user provided a non-uri local path?

And yes, WindowsUPath and PosixUPath could derive from LocalPath.

Reply:
> For your usecase do you just need to determine if a path is local, or if the user provided a non-uri local path?

For local paths, I'd rather not use fsspec. So, I want a way of determining whether I have local paths (essentially pathlib Paths). It doesn't matter to me whether the user omitted the protocol or put in `file://`.

Reply:
We don't cover that use case right now. And I think it didn't work this way previously either.

#### in `<=0.0.24`:
  
- with `UPath("file:///local/file.txt")` you received a `UPath` instance that managed file access through `fsspec`
- with `UPath("/local/file.txt")` you received a `PosixPath` instance.

#### in `>=0.1.0`:

- with `UPath("file:///local/file.txt")` you receive a `LocalPath` instance that manages file access through `fsspec`
- with `UPath("/local/file.txt")` you received a `PosixUPath` instance.

<br>

#### workaround
A current workaround to force pathlib handling might be:
```python
pth = UPath(...)

if isinstance(obj, LocalPath):
    obj = UPath(obj.path)  # workaround to force pathlib handling for "file://" uris
    assert isinstance(obj, (PosixUPath, WindowsUPath)) 
if isinstance(obj, (PosixUPath, WindowsUPath)):
    ...  # local.
else:
    ...  # cloud, etc.
```
but this isn't elegant at all.

<br>

#### options
we'd need a way to determine if file uris should be handled via fsspec or not. Maybe this should actually be an option that should go both ways?

```python
class LocalFileBackend(str, Enum):
    PATHLIB = "pathlib"
    FSSPEC = "fsspec"
    AUTO = "auto"

# --- via a new constructor kwarg ---

obj_a = UPath("/local/file.txt")
obj_b = UPath("file:///local/file.txt")
assert type(obj_a) is PosixUPath and type(obj_b) is LocalPath

obj_a = UPath("/local/file.txt", localfile_backend="pathlib")
obj_b = UPath("file:///local/file.txt", localfile_backend="pathlib")
assert type(obj_a) is PosixUPath and type(obj_b) is PosixUPath

obj_a = UPath("/local/file.txt", localfile_backend="fsspec")
obj_b = UPath("file:///local/file.txt", localfile_backend="fsspec")
assert type(obj_a) is LocalPath and type(obj_b) is LocalPath

# --- via a utility function ---

def ensure_local(pth: UPath, backend: LocalFileBackend) -> PosixUPath | WindowsUPath | LocalPath:
    """raise ValueError if not local, ensure backend if local"""
    ...

```

After writing, I would almost prefer an `ensure_local` helper function. But we can go either route.

What do you think?

Reply:
Interesting. Probably, I didn't run into this issue because I stripped the `file://` prefix.
Tbh for my purposes, `isinstance(obj, (PosixUPath, WindowsUPath))` is fine. If it were `isinstance(obj, (LocalUPath))` it would be easier. But no big deal.